feat: Added multi-log table support

sommmen opened the pull request:
@mo-esmp @followynne 

Dear friends,

I'm still waiting on #20, so i thought i'd put the work in.
Please review this PR.

![image](https://user-images.githubusercontent.com/8641495/224437825-129d0a87-4b66-4f0d-8d04-06c5c979ebbd.png)

It works like this:

IDataProvider was expanded with a `Name` Property*.
These can be implemented by the provider to give a comprehensible name.
For ex:

``` csharp
public class SqlServerDataProvider : IDataProvider
{
    public string Name => _options.ToDataProviderName("MsSQL");
...

public static class RelationalDbOptionsExtensions
{
    public static string ToDataProviderName(this RelationalDbOptions options, string providerName)
        => string.Join(".", "MSSQL", options.Schema, options.TableName);
}
```

Options (like `RelationalDbOptions`) are no longer added to DI, because they need to be associated with a single provider.
Providers now use factories that contain a single option instance:

``` csharp
((ISerilogUiOptionsBuilder)optionsBuilder).Services
    .AddScoped<IDataProvider, SqlServerDataProvider>(p => ActivatorUtilities.CreateInstance<SqlServerDataProvider>(p, relationProvider));
```

This allows registering multiple providers**, with their own options, and without needing some kind of string identifier to match options with providers***.

`AddSerilogUi` then adds an `AggregateDataProvider` that delegates to `IDataProvider`;

The middleware then has another endpoint 'api/keys' which returns a collection of key values for the dropdown in the UI.
It also listens for the 'key' query parameter value and will get the appropriate provider given a certain key.

``` csharp
/// <summary>
/// If there is only one data provider, this is it.
/// If there are multiple, this is the current data provider.
/// </summary>
public IDataProvider DataProvider => _dataProvider;

private readonly Dictionary<string, IDataProvider> _dataProviders = new Dictionary<string, IDataProvider>();

public void SwitchToProvider(string key)
    => _dataProvider = _dataProviders[key];

public IEnumerable<string> Keys => _dataProviders.Keys;
```

\* Names are made unique in case of collisions:
  ``` csharp
    // When providers with the same name are registered, we ensure uniqueness by generating a key
    // I.e. ["MSSQL.dbo.logs", "MSSQL.dbo.logs"] => ["MSSQL.dbo.logs[0]", "MSSQL.dbo.logs[1]"]
    for (var i = 0; i < grouped.Count; i++)
    {
        var dataProvider = grouped[i];
        _dataProviders.Add($"{name}[{i}]", dataProvider);
    }
  ```
** Some providers like ES / MongoDb use a singleton client and pass it into the dataprovider.  For those i disabled this feature (can only add one provider of that type).

*** In theory you could easily allow IDataProvider.Name to be confgurable so you can have custom names. I did not bother writing that as i think the use case would be minimal.

Feel free to make changes, scrap it all together or do anything to this code - i just want this feature one way or another ;).

TODO
 - Documentation in Readme (Maybe after some testing in production we can let this feature be fully known?)

followynne commented:
Hi @sommmen ,

Thanks for the pull request and the detailed description, from a quick look it seems great!

I'll try to give it a review asap, after I complete my work on integration tests and I review #63 (sorry but I've been quite busy... ).

In my opinion, it would be great to update the README (we can later think of dividing it in a small wiki).

If I merge soon my pr on the integration tests, would you be ok with creating some tests for the providers based on your multi-tables? If yes, I'll add you info later 😊

sommmen commented:
> Hi @sommmen ,
> 
> Thanks for the pull request and the detailed description, from a quick look it seems great!
> 
> I'll try to give it a review asap, after I complete my work on integration tests and I review #63 (sorry but I've been quite busy... ).
> 
> In my opinion, it would be great to update the README (we can later think of dividing it in a small wiki).
> 
> If I merge soon my pr on the integration tests, would you be ok with creating some tests for the providers based on your multi-tables? If yes, I'll add you info later 😊

Hiya,

> In my opinion, it would be great to update the README (we can later think of dividing it in a small wiki).

Sure, i'll write a readme if i'm cretain these changes will make it into the main branch and you guys don't want to go another route, i'll await your review.

> If I merge soon my pr on the integration tests, would you be ok with creating some tests for the providers based on your multi-tables?

Yeah, i'm not familiar with unit/integration testing a lot, but i'll figure it out.

mo-esmp commented:
@sommmen Thanks for the PR, I'll try to review it and give my feedback. I guess it takes time to merge the PR because of other PRs and implementing tests for it.

sommmen commented:
> @sommmen Thanks for the PR, I'll try to review it and give my feedback. I guess it takes time to merge the PR because of other PRs and implementing tests for it.

Sure thing, just as long as the ball is rolling 😄

sommmen commented:
@followynne 

I merged all changes from dev into this branch and fixed the registration tests.
I changed `GetRequiredService` to `GetService` so that the service provider does not throw an exception, and the test conditions for .NotNull would ran (effect is the same, but better error message in the test)

I also had to create a scope for the service provider because dataproviders are generally scoped and not singletons*

``` csharp
var serviceProvider = serviceCollection.BuildServiceProvider();
using var scope = serviceProvider.CreateScope();

var provider = scope.ServiceProvider.GetService<IDataProvider>();
provider.Should().NotBeNull().And.BeOfType<SqlServerDataProvider>();
```

Since options are no longer registered in the service provider, i'm fetching the private fields via reflection**:

``` csharp
var optionsField = typeof(SqlServerDataProvider)
                       .GetField("_options", BindingFlags.NonPublic | BindingFlags.Instance)
                   ?? throw new InvalidOperationException("_options field is missing.");
var options = (RelationalDbOptions)optionsField.GetValue(provider) ?? throw new InvalidOperationException("optionsField.GetValue(provider) returned null");
```

\* i'm wondering whether this worked before, because it should not have?
** An argument can be made that the options should no longer be unit-tested because of this, and instead an integration test should test this, but this was an easy fix for now that retains the intended behaviour.

All tests are green.

@mo-esmp 
I was not able to get to serilog-ui from the demo app.
I created an account Sample1@sample1.nl (same pw) and that seemed to work, but /serilog-ui gave a 'not sufficient permissions'.
Context.User seemed to be empty in the custom auth filter, so i'm assuming no token cookie or something was set from the /login page?

Could also have broken something whilst merging. 
Merging was pretty straight forward, but a lot was auto-merged by git.

Please advise :)

mo-esmp commented:
@sommmen in the `Startup` class and in `AddAuthentication` method comment default JWT authentication scheme (line 103)

followynne commented on src/Serilog.Ui.Core/IDataProvider.cs:
```diff
@@ -27,5 +27,13 @@ public interface IDataProvider
-}
\ No newline at end of file
+}
+namespace Serilog.Ui.Core.Services
```
(To be cleaned up)

followynne commented on src/Serilog.Ui.ElasticSearchProvider/ElasticSearchDbDataProvider.cs:
```diff
@@ -10,13 +10,16 @@ namespace Serilog.Ui.ElasticSearchProvider
         {
             _client = client ?? throw new ArgumentNullException(nameof(client));
             _options = options ?? throw new ArgumentNullException(nameof(options));
+            Name = string.Join(".", "ES", _options.IndexName);
```
I would move this set from ctor to the property initialization itself, to uniform it like in MsSql...

sommmen replied:
Yes, due to the non-static nature, `=>` must be used (defines only a GET method).

Take note that this should be used sparingly, because `{get;set;} = "hi"` will only instantiate the property once.
Defining a get method only will result in the result being return multiple times (more allocations etc).

In this case its okay because the property is not read much, `AggregateDataProvider` caches the keys.

See:
https://sharplab.io/#v2:D4AQTAjAsAUCDMACciDCiDetaN3/uCyEADIgHICGAtgKaIC8AfMSQHQBSA9gJYB2ACgBEbIQBpEQgKIBlcZLBCAlAG5sMHHiIgALIgCyApZk24AvrDNA

followynne commented on src/Serilog.Ui.MongoDbProvider/MongoDbDataProvider.cs:
```diff
@@ -9,14 +9,20 @@ namespace Serilog.Ui.MongoDbProvider
+            _collection = client.GetDatabase(options.DatabaseName)
+                .GetCollection<MongoDbLogModel>(options.CollectionName);
+
+            Name = string.Join(".", "MongoDb", options.DatabaseName, options.CollectionName);
```
same as per https://github.com/mo-esmp/serilog-ui/pull/66/files#r1134480582

followynne commented on src/Serilog.Ui.MsSqlServerProvider/Extensions/SerilogUiOptionBuilderExtensions.cs:
```diff
@@ -1,47 +1,49 @@
+            };
+
+            ((ISerilogUiOptionsBuilder)optionsBuilder).Services
+                .AddScoped<IDataProvider, SqlServerDataProvider>(p => ActivatorUtilities.CreateInstance<SqlServerDataProvider>(p, relationProvider));
```
As a curiosity, why are you using ActivatorUtilities instead of doing a simple 
`p => new SqlServerDataProvider(relationProvider)?`

sommmen replied:
`ActivatorUtilities.CreateInstance` will use the service provider to instantiate services, along with any services you provide.
This is nice because if in the future a dependency service is added to the data provider, this can be registered in the container (so you can scope it accordingly) and it will automatically be injected into the data provider without requiring additional code.

followynne commented on src/Serilog.Ui.MySqlProvider/Extensions/SerilogUiOptionBuilderExtensions.cs:
```diff
@@ -1,42 +1,43 @@
+            };
+            
+            ((ISerilogUiOptionsBuilder)optionsBuilder).Services
+                .AddScoped<IDataProvider, MySqlDataProvider>(p => ActivatorUtilities.CreateInstance<MySqlDataProvider>(p, relationProvider));
```
same as per https://github.com/mo-esmp/serilog-ui/pull/66/files#r1134483389

followynne commented on src/Serilog.Ui.PostgreSqlProvider/Extensions/SerilogUiOptionBuilderExtensions.cs:
```diff
@@ -1,47 +1,47 @@
+            };
+
+            ((ISerilogUiOptionsBuilder)optionsBuilder).Services
+                .AddScoped<IDataProvider, PostgresDataProvider>(p => ActivatorUtilities.CreateInstance<PostgresDataProvider>(p, relationProvider));
```
same as per https://github.com/mo-esmp/serilog-ui/pull/66/files#r1134483389

followynne commented on src/Serilog.Ui.Web/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -30,6 +33,12 @@ public static IServiceCollection AddSerilogUi(this IServiceCollection services,
 
+            services.TryAddScoped(typeof(AggregateDataProvider), p =>
+            {
+                var providers = p.GetServices<IDataProvider>();
```
Is this manual registration required? I was thinking that .NET should already pass all the registered IEnumerable<IDataProvider> by itself, when you add AggregateDataProvider as a service...

followynne commented on src/Serilog.Ui.Web/SerilogUiMiddleware.cs:
```diff
@@ -170,7 +215,14 @@ private async Task<string> FetchLogsAsync(HttpContext httpContext)
-            var provider = httpContext.RequestServices.GetService<IDataProvider>();
+            var provider = httpContext.RequestServices.GetService<AggregateDataProvider>();
+
+            string key = keyStr;
```
this variable assignation can be removed...

followynne commented on src/Serilog.Ui.Core/AggregateDataProvider.cs:
```diff
@@ -0,0 +1,72 @@
+    /// <summary>
+    /// Aggregates multiple <see cref="IDataProvider"/> into one instance.
+    /// </summary>
+    public class AggregateDataProvider : IDataProvider
```
An idea about this one: 
probably we could change Aggregate[...] from being an IDataProvider implementation to being a separate interface with Strategy patternsince it doesn't need to implement the IDataProvider interface...
implementation would not really change, it would simply get the usual strategy ptn methods...
(probably later we could add a Action<?> option, to let the user select the provider by a delegate - not really sure it would make sense 😆 )

sommmen replied:
> probably we could change Aggregate[...] from being an IDataProvider implementation to being a separate interface with Strategy patternsince it doesn't need to implement the IDataProvider interface...
implementation would not really change, it would simply get the usual strategy ptn methods...

Yeah that makes sense.
If its okay with you i'd like to merge this into the repo and refactor this later.

It will be something like:

``` csharp
public class DataProviderContext {
     public prop DataProvider { get; }
     
     public SetDataProvider(string key) { ... }
}
```

With an interface.

followynne replied:
IMHO agreed! 😃

sommmen commented:
@followynne Thanks for the review and spotting an error or 2, i've resolved most comments let me know if there is more.

> thanks a lot for the latest comment - agreed on the scope creation, I suppose that the framework in the tests either created a Singleton instance of the provider or it created a scope on the fly (I think the 1st).

That makes sense.

> About the options fields that are failing in MsSql, MYSql, Postgres, you can remove those lines - you're 100% correct that they should not be tested anymore, as they're not DI-registered anymore.

Removed testing for options.

followynne reviewed (approved):
Awesome job - good to go in my opinion! 🥇

mo-esmp commented:
@sommmen a question regarding grouping data providers. If I have two log table in two different databases, does still it works fine?

sommmen commented:
> @sommmen a question regarding grouping data providers. If I have two log table in two different databases, does still it works fine?

It definitely should work, yes. 

One thing i can imagine is that the naming strategy i'm using now does not take the server into account, so i could imagine the following:

2 different databases (maybe even on 2 different servers), but with same schema and table name:

``` csharp
MsSql.AppLog.SeriLogs[0]
MsSql.AppLog.SeriLogs[1]
```

This is a good case for allowing the user to override the names, which could be a cool next step.
We could do something fancy like extracting the connection string and parse it to generate unique names but that seems like overengineering imo.

If you have different providers - this would not matter;

``` csharp
MsSql.AppLog.SeriLogs
MySql.AppLog.SeriLogs
```

mo-esmp commented:
Generating names from connection strings should not be difficult `ServerName-DbName-TableName` and avoids name conflict.

sommmen commented:
> Generating names from connection strings should not be difficult `ServerName-DbName-TableName` and avoids name conflict.

Hiya - I have this change ready locally. 
It was not hard because of .net `DbConnectionStringBuilder` which'd handle the different kinds of connection strings per provider.

``` csharp
public static string ToDataProviderName(this RelationalDbOptions options, string providerName,
    string serverName)
    => string.Join(".", providerName, serverName, options.Schema, options.TableName);

public string Name => _options.ToDataProviderName("NPGSQL", new NpgsqlConnectionStringBuilder(_options.ConnectionString).Host);
```

But i had some concerns regarding sercurity,
Generally connection strings should remain a secret, so exposing this to the UI may not be a good option.

I think naming strategy as-is will provide enough uniqueness, atleast for now.

For me i have several apps running on the same database/server with different log tables.
I doubt someone will run 1 dashboard for multiple apps across multiple servers - is this a use-case for you?

The current naming strategy would cover those cases.

Let me know if you want the server to be part of the name regardless, or otherwise you can merge the changes in at your leasure.

mo-esmp commented:
Well, That could be an option to check all logs for different servers from one dashboard instead of installing SerilogUI and checking them one by one. For now, it's ok but later we can add this feature also.

sommmen commented:
@mo-esmp Hiya, could you release this feature?

sommmen commented:
> The nuget package v2.4.0 already published!

You're right, somehow the changes did not show for me last time i tried it. Maybe vs caching or something.
Thanks!

Bluejanis commented:
Being able to override the default name would be nice to have.  Preferably, when you configure the sql server, so another optional parameter in `UseSqlServer` would be nice for this.

followynne commented:
> Being able to override the default name would be nice to have.  Preferably, when you configure the sql server, so another optional parameter in `UseSqlServer` would be nice for this.

Hello @Bluejanis,

We can try to make it fit in the upcoming v3 release.

Can I ask you to open a separate issue, with the precise request, instead of writing it in a closed PR? It will help us track your feature better. 😊
